{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sandra/work/Bioinformatics/Soybeans\n"
     ]
    }
   ],
   "source": [
    "cd /Users/sandra/work/Bioinformatics/Soybeans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/sandra/work/Bioinformatics/Soybeans'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'> [Glycine max] Forrest B100B10 BAC sequence, Rhg4 \\nCCCTCGCGAGTTGGTTCAGCTGCTGCCTGAGGCTGGACGACCTCGCGG'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('Forrest B100B10 BAC sequence Rhg4.txt', 'r') as forrest_file:\n",
    "    forrest_seq = forrest_file.read()\n",
    "\n",
    "forrest_seq[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'> [Glycine max] Forrest B100B10 BAC sequence, Rhg4CCCTCGCGAGTTGGTTCAGCTGCTGCCTGAGGCTGGACGACCTCGCGGAGTTCTACCGGCAGTGCAAATCCGTCGGCATCCAGGAAACCAGCAGCGGCTATCCGCGCATCCATGCCCCCGAACTGCAGGAGTGGGGAGGCACGATGGCCGCTTTGGTCGACCTGCAGGGGGGGGGGGCGCTGAGGTCTGCCTCGTGAAGAAGGTGTTGCTGACTCATACCAGGCCTGAATCGCCCCATCATCCAGCCAGA'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('Forrest B100B10 BAC sequence Rhg4.txt', 'r') as forrest_file:\n",
    "    forrest_lines = ''\n",
    "    file_line = forrest_file.readline().rstrip()\n",
    "    # forrest_lines += file_line\n",
    "\n",
    "    while file_line:\n",
    "        forrest_lines += file_line\n",
    "        file_line = forrest_file.readline().rstrip()\n",
    "\n",
    "forrest_lines[:300]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The file containing the Forrest sequence I am working with starts with the FASTA format header. This header doesn't help much in downstreadm analysis, so I need to find index where actual sequence starts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "forrest_lines.find('Rhg4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'R'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "forrest_lines[46]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Rhg4'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "forrest_lines[46:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since Python starts index at 0, Rhg4 (name of the region/locus) is indices 46-49. Actual Forrest sequence starts at index 50."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCCTC'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "forrest_lines[50:55]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I am proceeding to searching for presence of genes in the ranges of alignment. I am starting with the ranges of alignment for the soybean cultivar Williams chromosomes 8 and 5, since when running Nucleotide BLAST, they gave the highest total score and E value. Using the Forrest Rhg4 sequence, I am adjusting the boundaries given by BLAST for the FASTA format header."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### When I performed the alignment using Nucleotide BLAST, the first range of alignment was 1036 through 40810 in my input Forrest sequence. Adjusting for the way Python does indexing, this interval becomes 1036-40811 (we want Python to go through base 40810). Length of FASTA format header is 50."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_codon = 'ATG'\n",
    "sequence = forrest_lines[1036+50:40811]\n",
    "sequence.find(start_codon)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "indices of stop codons in the 1037-40810 range of alignment: [16, 6, 59]\n"
     ]
    }
   ],
   "source": [
    "stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "sequence = forrest_lines[1036+50:40811]\n",
    "\n",
    "stop_codon_indices = []\n",
    "for codon in stop_codons:\n",
    "    index = sequence.find(codon)\n",
    "    stop_codon_indices.append(index)\n",
    "\n",
    "print(f'indices of stop codons in the 1037-40810 range of alignment: {stop_codon_indices}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stop indices reversed: [57, 98, 266]\n",
      "stop codon: CAA\n",
      "AAC\n"
     ]
    }
   ],
   "source": [
    "# trying to reverse the stop codons\n",
    "\n",
    "stop_codon_indices = []\n",
    "for codon in stop_codons:\n",
    "    index = sequence.find(codon[::-1])\n",
    "    stop_codon_indices.append(index)\n",
    "\n",
    "stop_codon = sequence[263:266][::-1]\n",
    "print(f'stop indices reversed: {stop_codon_indices}')\n",
    "print(f'stop codon: {stop_codon}')\n",
    "\n",
    "print(sequence[263:266])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 733\n",
      "first 10 indices of start codon: [58, 146, 211, 306, 381, 522, 689, 1011, 1074, 1090]\n"
     ]
    }
   ],
   "source": [
    "# loop over the sequence when looking for start codon\n",
    "\n",
    "start_codon = 'ATG'\n",
    "start_codon_indices = []\n",
    "sequence = forrest_lines[1036+50:40811]\n",
    "\n",
    "index = sequence.find(start_codon)\n",
    "while index >= 0:\n",
    "    start_codon_indices.append(index)\n",
    "    index = sequence.find(start_codon, index+len(start_codon))\n",
    "\n",
    "print(f'number of possible start codons: {len(start_codon_indices)}')\n",
    "print(f'first 10 indices of start codon: {start_codon_indices[:10]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of TATA boxes: 51\n",
      "first 10 indices of TATA boxes: [204, 376, 432, 1998, 2031, 3615, 3944, 4539, 5088, 5185]\n"
     ]
    }
   ],
   "source": [
    "# finding TATA boxes\n",
    "tata_box = 'TATAAA'\n",
    "sequence = forrest_lines[1036+50:40811]\n",
    "tata_indices = []\n",
    "\n",
    "index = sequence.find(tata_box)\n",
    "while index >= 0:\n",
    "    tata_indices.append(index)\n",
    "    index = sequence.find(tata_box, index + len(tata_box))\n",
    "\n",
    "print(f'number of TATA boxes: {len(tata_indices)}')\n",
    "print(f'first 10 indices of TATA boxes: {tata_indices[:10]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since TATA boxes are 25-35 base pairs upstream of the transcription start site in most eukaryotic genes, we can rule out codons at indices 58, 146, 211 as start codons and presence of genes. Starting gene search from index 1419 (start of sequence + index of TATA box + length of the TATA box + 25 bp downstream of the TATA box)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 729\n"
     ]
    }
   ],
   "source": [
    "# start codons starting from 25 bp downstream of the TATA box\n",
    "codon = 'ATG'\n",
    "sequence = forrest_lines[1419:40811]\n",
    "\n",
    "poss_start_codons = []\n",
    "index = sequence.find(codon)\n",
    "while index >= 0:\n",
    "    poss_start_codons.append(index)\n",
    "    index = sequence.find(codon, index + len(codon))\n",
    "\n",
    "print(f'number of possible start codons: {len(poss_start_codons)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible stop codons: 2528\n"
     ]
    }
   ],
   "source": [
    "# stop codons starting from 35 bp downstream of the TATA box (TATA box 25-35 bp upstream of the gene)\n",
    "stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "sequence = forrest_lines[1419+10:40811]\n",
    "\n",
    "poss_stop_codons = []\n",
    "for codon in stop_codons:\n",
    "    index = sequence.find(codon)\n",
    "    while index >= 0:\n",
    "        poss_stop_codons.append(index)\n",
    "        index = sequence.find(codon, index + len(codon))\n",
    "\n",
    "print(f'number of possible stop codons: {len(poss_stop_codons)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We got possibile start and stop codons. Let's calculate the difference between indices for each pair to get a sense of presence of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "possible start codons[:10]: [48, 189, 356, 678, 741, 757, 834, 885, 933, 1014]\n",
      "possible stop codons[:10]: [10, 27, 35, 65, 91, 102, 171, 177, 262, 423]\n"
     ]
    }
   ],
   "source": [
    "print(f'possible start codons[:10]: {poss_start_codons[:10]}')\n",
    "print(f'possible stop codons[:10]: {poss_stop_codons[:10]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible genes: 899469\n"
     ]
    }
   ],
   "source": [
    "# iterating through the list of possible start codons\n",
    "poss_genes = []\n",
    "for start in poss_start_codons:\n",
    "    for stop in poss_stop_codons:\n",
    "        if stop > start:\n",
    "            length = stop - start\n",
    "            # print(f'stop codon - start codon: {length}')\n",
    "            poss_genes.append(length)\n",
    "\n",
    "# expecting to have 729 possible genes (same as number of possible start codons), but check\n",
    "print(f'number of possible genes: {len(poss_genes)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of possible I got is greater than expected (there shouldn't be more genes than start codons). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Changing how I'm searching for gene presence. Using dictionary instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first 5 possible start codons: [48, 189, 356, 678, 741]\n"
     ]
    }
   ],
   "source": [
    "print(f'first 5 possible start codons: {poss_start_codons[:5]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible genes: 38216\n",
      "first 5 possible genes: [(17, (39301, 39318)), (43, (39244, 39287)), (54, (39301, 39355)), (123, (37961, 38084)), (129, (36995, 37124))]\n"
     ]
    }
   ],
   "source": [
    "poss_genes_dict = {}\n",
    "for start in poss_start_codons:\n",
    "    for stop in poss_stop_codons:\n",
    "        if stop > start:\n",
    "            poss_gene_len = stop - start\n",
    "            poss_genes_dict[poss_gene_len] = (start, stop)\n",
    "\n",
    "print(f'number of possible genes: {len(poss_genes_dict)}')\n",
    "# print([poss_genes_dict[k] for k in list(poss_genes_dict.keys())[:5]])\n",
    "print(f'first 5 possible genes: {list(poss_genes_dict.items())[:5]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "possible genes: 686\n"
     ]
    }
   ],
   "source": [
    "# different approach: use while loop or change for loop\n",
    "\n",
    "poss_genes_dict = {}\n",
    "i_start = 0\n",
    "i_stop = 0\n",
    "while i_start < len(poss_start_codons):\n",
    "    if i_stop >= i_start:\n",
    "        length = poss_stop_codons[i_stop] - poss_start_codons[i_start]\n",
    "        poss_genes_dict[length] = (poss_start_codons[i_start], poss_stop_codons[i_stop])\n",
    "        i_start += 1\n",
    "        i_stop += 1\n",
    "    else:\n",
    "        i_stop += 1\n",
    "\n",
    "print(f'possible genes: {len(poss_genes_dict)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We have a reasonable number of possible genes in this region of the Forrest sequence. It is likely that the other 30 possible start codons are located such that their indices are bigger than the indices of the corresponding possible stop codons. Putting this into a function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_possible_genes(poss_start_codons, poss_stop_codons):\n",
    "    \"\"\"Function takes list of possible start codons and list of possible stop codons.\n",
    "    It returns dictionary with the number of possible genes, where keys are the length of each possible gene\n",
    "    and values are the start, stop indices where the gene is.\"\"\"\n",
    "    poss_genes_dict = {}\n",
    "    i_start = 0\n",
    "    i_stop = 0\n",
    "    while i_start < len(poss_start_codons):\n",
    "        if i_stop > i_start:\n",
    "            length = poss_stop_codons[i_stop] - poss_start_codons[i_start]\n",
    "            poss_genes_dict[length] = (poss_start_codons[i_start], poss_stop_codons[i_stop])\n",
    "            i_start += 1\n",
    "            i_stop += 1\n",
    "        else:\n",
    "            i_stop += 1\n",
    "\n",
    "        return poss_genes_dict\n",
    "\n",
    "get_possible_genes (poss_start_codons, poss_stop_codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Troubleshooting code for finding possible genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'> [Glycine max] Forrest B100B10 BAC sequence, Rhg4CCCTCGCGAGTTGGTTCAGCTGCTGCCTGAGGCTGGACGACCTCGCGGAGTTCTACCGGCAGTGCAAATCCGTCGGCATCCAGGAAACCAGCAGCGGCTATCCGCGCATCCATGCCCCCGAACTGCAGGAGTGGGGAGGCACGATGGCCGCTTTGGTCGACCTGCAGGGGGGGGGGGCGCTGAGGTCTGCCTCGTGAAGAAGGTGTTGCTGACTCATACCAGGCCTGAATCGCCCCATCATCCAGCCAGA'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('Forrest B100B10 BAC sequence Rhg4.txt', 'r') as forrest_file:\n",
    "    forrest_seq = ''\n",
    "    file_line = forrest_file.readline().rstrip()\n",
    "    # forrest_lines += file_line\n",
    "\n",
    "    while file_line:\n",
    "        forrest_seq += file_line\n",
    "        file_line = forrest_file.readline().rstrip()\n",
    "\n",
    "forrest_seq[:300]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 729\n"
     ]
    }
   ],
   "source": [
    "def get_codon_indices(codon, sequence, start_point):\n",
    "    poss_indices = []\n",
    "    index = sequence.find(codon, start_point)\n",
    "    while index >= 0:\n",
    "        poss_indices.append(index)\n",
    "        index = sequence.find(codon, index+len(codon))\n",
    "    \n",
    "    return poss_indices\n",
    "\n",
    "start_codon = 'ATG'\n",
    "sequence = forrest_lines[:40811]\n",
    "start_point = 1419\n",
    "poss_start_codon_ind = get_codon_indices(start_codon, sequence, start_point)\n",
    "print(f'number of possible start codons: {len(poss_start_codon_ind)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible stop codons: 833\n"
     ]
    }
   ],
   "source": [
    "stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "start_point = 1419\n",
    "\n",
    "poss_stop_codon_ind = []\n",
    "for codon in stop_codons:\n",
    "    poss_stop_codon_ind = get_codon_indices(codon, sequence, start_point)\n",
    "\n",
    "print(f'number of possible stop codons: {len(poss_stop_codon_ind)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "indices of first 5 possible start codons: [1467, 1608, 1775, 2097, 2160]\n",
      "indices of first 5 possible stop codons: [1468, 1486, 1795, 2079, 2203]\n"
     ]
    }
   ],
   "source": [
    "print(f'indices of first 5 possible start codons: {poss_start_codon_ind[:5]}')\n",
    "print(f'indices of first 5 possible stop codons: {poss_stop_codon_ind[:5]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 729\n",
      "if branch: poss_start_codons[i_start]:1467, poss_stop_codons[i_stop]:1468\n",
      "length: 1\n",
      "number of possible genes: 1\n",
      "possible genes: {(1467, 1468): 1}\n"
     ]
    }
   ],
   "source": [
    "# let's run the get_possible_genes function again\n",
    "\n",
    "def get_possible_genes(poss_start_codons, poss_stop_codons):\n",
    "    \"\"\"Function takes list of possible start codons and list of possible stop codons.\n",
    "    It returns dictionary with the number of possible genes, where keys are the length of each possible gene\n",
    "    and values are the start, stop indices where the gene is.\"\"\"\n",
    "    print(f'number of possible start codons: {len(poss_start_codons)}')\n",
    "    poss_genes_dict = {}\n",
    "    # i_start = 0\n",
    "    i_stop = 0\n",
    "    for i_start in range(len(poss_start_codons)):\n",
    "        if poss_stop_codons[i_stop] > poss_start_codons[i_start]:\n",
    "            print(f'if branch: poss_start_codons[i_start]:{poss_start_codons[i_start]}, poss_stop_codons[i_stop]:{poss_stop_codons[i_stop]}')\n",
    "            length = poss_stop_codons[i_stop] - poss_start_codons[i_start]\n",
    "            print(f'length: {length}')\n",
    "            poss_genes_dict[(poss_start_codons[i_start], poss_stop_codons[i_stop])] = length\n",
    "            i_start += 1\n",
    "            i_stop += 1\n",
    "        else:\n",
    "            print(f'else branch: poss_start_codons[i_start]:{poss_start_codons[i_start]}, poss_stop_codons[i_stop]:{poss_stop_codons[i_stop]}')\n",
    "            i_stop += 1\n",
    "\n",
    "        return poss_genes_dict\n",
    "\n",
    "poss_genes = get_possible_genes(poss_start_codon_ind, poss_stop_codon_ind)\n",
    "print(f'number of possible genes: {len(poss_genes)}')\n",
    "print(f'possible genes: {poss_genes}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Changing to finding indices possible stop codons using each stop codon separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 729\n"
     ]
    }
   ],
   "source": [
    "def find_possible_codons(codon, sequence):\n",
    "    poss_codons = []\n",
    "    index = sequence.find(codon)\n",
    "    while index >= 0:\n",
    "        poss_codons.append(index)\n",
    "        index = sequence.find(codon, index + len(codon))\n",
    "    return poss_codons\n",
    "\n",
    "start_codon = 'ATG'\n",
    "sequence = forrest_lines[1419:40811]\n",
    "poss_start_codons = find_possible_codons(start_codon, sequence)\n",
    "print(f'number of possible start codons: {len(poss_start_codons)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible start codons: 729\n"
     ]
    }
   ],
   "source": [
    "def find_possible_codon_indices(codon, sequence):\n",
    "    poss_codon_ind = []\n",
    "    index = sequence.find(codon)\n",
    "    while index >= 0:\n",
    "        poss_codon_ind.append(index)\n",
    "        index = sequence.find(codon, index + len(codon))\n",
    "    return poss_codon_ind\n",
    "\n",
    "start_codon = 'ATG'\n",
    "start_point = 1419\n",
    "end_point = 40811\n",
    "sequence = forrest_lines[start_point:end_point]\n",
    "poss_start_codon_ind = find_possible_codon_indices(start_codon, sequence)\n",
    "print(f'number of possible start codons: {len(poss_start_codon_ind)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### When looking for indices of possible stop codons, I chose to not take mismatches like TAA to GAA into account. In addition to T (Thymine) being a because biologically speaking a lot would have to go wrong during DNA replication for such a mutation to occur. T (thymine) is a pyrimidine while G (guanine) is a purine. If in the parent DNA has T on one strand, it would have A (adenine, purine like G) on the other strand. If DNA Polymerase tried pairing G with A, this would distort the newly forming DNA molecule. Also, the structure of DNA Polymerase III is such that one of its domains (parts) corrects mistakes made during replication (like pairing C (cytosine) with A instead of T with A)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### One of the problems with the running of get_possible_genes function resulted from putting indices for all three stop codons into one single output list. Changing to using one output list for each of the three stop codons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible TAA stop codons: 1163\n",
      "number of possible TAG stop codons: 533\n",
      "number of possible TGA stop codons: 833\n"
     ]
    }
   ],
   "source": [
    "stop_codon = 'TAA'\n",
    "poss_taa_stop_codons = find_possible_codons(stop_codon, sequence)\n",
    "print(f'number of possible {stop_codon} stop codons: {len(poss_taa_stop_codons)}')\n",
    "stop_codon = 'TAG'\n",
    "poss_tag_stop_codons = find_possible_codons(stop_codon, sequence)\n",
    "print(f'number of possible {stop_codon} stop codons: {len(poss_tag_stop_codons)}')\n",
    "stop_codon = 'TGA'\n",
    "poss_tga_stop_codons = find_possible_codons(stop_codon, sequence)\n",
    "print(f'number of possible {stop_codon} stop codons: {len(poss_tga_stop_codons)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have the number of possible stop codon indices for each of the three stop codons. Back to finding possible genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible genes with TAA stop codons: 25\n"
     ]
    }
   ],
   "source": [
    "def find_poss_genes(poss_start_codons, poss_stop_codons):\n",
    "    # print(f'first 5 possible start codons: {poss_start_codons[:5]}, first 5 possible stop codons: {poss_stop_codons[:5]}')\n",
    "    poss_genes = {}\n",
    "    i_start = 0\n",
    "    i_stop = 0\n",
    "\n",
    "    while i_start < len(poss_start_codons) and i_stop < len(poss_stop_codons):\n",
    "        if poss_stop_codons[i_stop] > poss_start_codons[i_start]:\n",
    "            gene_length = poss_stop_codons[i_stop] - poss_start_codons[i_start]\n",
    "            # prokaryotes have a typical gene length of 1000 base pairs and each codon is encoded by 3 bases\n",
    "            if gene_length >= 1000:\n",
    "                poss_genes[(poss_stop_codons[i_stop], poss_start_codons[i_start])] = gene_length\n",
    "            i_start += 1\n",
    "            i_stop += 1\n",
    "            continue\n",
    "        else:\n",
    "            i_stop += 1\n",
    "    \n",
    "    return poss_genes\n",
    "\n",
    "taa_genes = find_poss_genes(poss_start_codons, poss_taa_stop_codons)\n",
    "print(f'number of possible genes with TAA stop codons: {len(taa_genes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of possible genes with TAG stop codons: 477\n",
      "number of possible genes with TGA stop codons: 0\n"
     ]
    }
   ],
   "source": [
    "tag_genes = find_poss_genes(poss_start_codons, poss_tag_stop_codons)\n",
    "print(f'number of possible genes with TAG stop codons: {len(tag_genes)}')\n",
    "\n",
    "tga_genes = find_poss_genes(poss_start_codons, poss_tga_stop_codons)\n",
    "print(f'number of possible genes with TGA stop codons: {len(tga_genes)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have a working algorithm, but the way everything is written I need to run it for each range of alignment. It would be nice to have a way for the algorithm to run through the Forrest sequence without me having to manually move it from one range of alignment to another."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plotting frequency of genes in the 1037-40810 range of alignment for TAA and TAG stop codons (this range of alignment did not have any possible genes that started at or after base pair 1419 for TGA stop codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "taa_genes = find_poss_genes(poss_start_codons, poss_taa_stop_codons)\n",
    "# print(f'possible genes with TAA as the stop codon: {taa_genes}')\n",
    "tag_genes = find_poss_genes(poss_start_codons, poss_tag_stop_codons)\n",
    "# print(f'possible genes with TAG as the stop codon: {tag_genes}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taa_gene_length: 25\n",
      "\n",
      "tag_gene_length: 477\n"
     ]
    }
   ],
   "source": [
    "taa_gene_length = []\n",
    "for index_pairs, length in taa_genes.items():\n",
    "    taa_gene_length.append(length)\n",
    "\n",
    "print(f'taa_gene_length: {len(taa_gene_length)}')\n",
    "print()\n",
    "\n",
    "tag_gene_length = []\n",
    "for index_pairs, length in tag_genes.items():\n",
    "    tag_gene_length.append(length)\n",
    "\n",
    "print(f'tag_gene_length: {len(tag_gene_length)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Count')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAWZklEQVR4nO3de5RsZX3m8e8DiIqIgLSMN2x0GC9xGUyOF4QxRtTBaCQqioxGEo2sLE3wMrngMqPMENfI6LhMdKKeZQiGELzgDcUoCAio3A4ICCLRUdTjjYMmEaNLRH/zx97tKZqqPn2aU7W7+/1+1urVu97al1+9Xf3Urreq3kpVIUlqx05DFyBJmi2DX5IaY/BLUmMMfklqjMEvSY3ZZegClmOfffap+fn5ocuQpDXl8ssvv6mq5ha3r4ngn5+fZ9OmTUOXIUlrSpKvj2t3qEeSGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1ZmrBn+SkJDcmuWak7Y1JvpTk6iQfSrLntI4vSRpvmmf8JwOHLWo7G3h4VT0C+Gfg1VM8viRpjKkFf1VdAPxgUdtZVXVrf/Fi4H7TOr4kabwhP7n7IuC9k65McgxwDMB+++03q5okrWLzx505dAkzd8MbnrbD9znIi7tJXgPcCpw6aZ2q2lhVG6pqw9zc7aaakCSt0MzP+JMcDTwdOLT83kdJmrmZBn+Sw4A/B36jqn48y2NLkjrTfDvnacBFwIOTbE7yYuBtwN2Bs5NcmeQd0zq+JGm8qZ3xV9VRY5r/dlrHkyQtj5/claTGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGTC34k5yU5MYk14y07Z3k7CRf7n/vNa3jS5LGm+YZ/8nAYYvajgPOqaoDgHP6y5KkGZpa8FfVBcAPFjUfDry7X3438DvTOr4kabxZj/HvW1XfAeh/32vSikmOSbIpyaYtW7bMrEBJWu9W7Yu7VbWxqjZU1Ya5ubmhy5GkdWPWwf+9JPcG6H/fOOPjS1LzZh38ZwBH98tHAx+Z8fElqXnTfDvnacBFwIOTbE7yYuANwJOTfBl4cn9ZkjRDu0xrx1V11ISrDp3WMSVJ27ZqX9yVJE2HwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjRkk+JO8Msm1Sa5JclqSuwxRhyS1aObBn+S+wLHAhqp6OLAz8LxZ1yFJrRpqqGcX4K5JdgF2A749UB2S1JyZB39VfQt4E/AN4DvAv1XVWYvXS3JMkk1JNm3ZsmXWZUrSujXEUM9ewOHA/sB9gLslecHi9apqY1VtqKoNc3Nzsy5TktatIYZ6ngR8raq2VNXPgA8CjxugDklq0hDB/w3gsUl2SxLgUOC6AeqQpCYNMcZ/CXA6cAXwhb6GjbOuQ5JatcsQB62q1wGvG+LYktQ6P7krSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1ZlnBn+Tg5bRJkla/5Z7xv3WZbZKkVW7J2TmTHET3JSlzSV41ctUedF+SLklaY7Y1LfOuwO79encfaf8hcMS0ipIkTc+SwV9V5wPnJzm5qr4+o5okSVO03C9iuXOSjcD86DZV9cRpFCVJmp7lBv/7gXcA7wJ+Pr1yJEnTttzgv7Wq3j7VSiRJM7Hct3N+NMlLk9w7yd4LP1OtTJI0Fcs94z+6//2nI20FPHDHliNJmrZlBX9V7T/tQiRJs7Gs4E/ywnHtVfX3O7YcSdK0LXeo51Ejy3cBDgWuAAx+SVpjljvU88ejl5PcAzhlKhVJkqZqpdMy/xg4YKUHTbJnktOTfCnJdf2cQJKkGVjuGP9H6d7FA93kbA8F3ncHjvtXwCeq6ogkuwK73YF9SZK2w3LH+N80snwr8PWq2rySAybZA3g88HsAVXULcMtK9iVJ2n7LHeM/P8m+bH2R98t34JgPBLYAf5fkV4HLgZdX1b+PrpTkGOAYgP322+8OHE5an+aPO3PoErRGLfcbuJ4LXAo8B3gucEmSlU7LvAvwa8Dbq+qRwL8Dxy1eqao2VtWGqtowNze3wkNJkhZb7lDPa4BHVdWNAEnmgE8Bp6/gmJuBzVV1SX/5dMYEvyRpOpb7rp6dFkK/9/3t2PY2quq7wDeTPLhvOhT44kr2JUnafss94/9Ekk8Cp/WXjwQ+fgeO+8fAqf07er4K/P4d2JckaTts6zt3/yOwb1X9aZJnAYcAAS4CTl3pQavqSmDDSreXJK3ctoZr3gLcDFBVH6yqV1XVK+nO9t8y3dIkSdOwreCfr6qrFzdW1Sa6r2GUJK0x2wr+uyxx3V13ZCGSpNnYVvBfluQlixuTvJjug1eSpDVmW+/qeQXwoSTPZ2vQbwB2BZ45xbokSVOyZPBX1feAxyX5TeDhffOZVXXu1CuTJE3FcufqOQ84b8q1SJJmYKXz8UuS1iiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwYL/iQ7J/l8ko8NVYMktWjIM/6XA9cNeHxJatIgwZ/kfsDTgHcNcXxJatlQZ/xvAf4M+MWkFZIck2RTkk1btmyZWWGStN7NPPiTPB24saouX2q9qtpYVRuqasPc3NyMqpOk9W+IM/6DgWckuQF4D/DEJP8wQB2S1KSZB39Vvbqq7ldV88DzgHOr6gWzrkOSWuX7+CWpMbsMefCq+jTw6SFrkKTWeMYvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1ZtC5emZh/rgzhy5h5m54w9OGLmHmWvw7SyvlGb8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjZh78Se6f5Lwk1yW5NsnLZ12DJLVsiGmZbwX+W1VdkeTuwOVJzq6qLw5QiyQ1Z+Zn/FX1naq6ol++GbgOuO+s65CkVg06xp9kHngkcMmY645JsinJpi1btsy8NklarwYL/iS7Ax8AXlFVP1x8fVVtrKoNVbVhbm5u9gVK0jo1SPAnuRNd6J9aVR8cogZJatUQ7+oJ8LfAdVX15lkfX5JaN8QZ/8HA7wJPTHJl//NbA9QhSU2a+ds5q+ozQGZ9XElSx0/uSlJjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxQ3znrqZs/rgzhy5B0irmGb8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjBgn+JIcluT7JV5IcN0QNktSqmQd/kp2B/ws8FXgYcFSSh826Dklq1RBn/I8GvlJVX62qW4D3AIcPUIckNWmIL2K5L/DNkcubgccsXinJMcAx/cUfJbl+O4+zD3DTiiqcPWvd8dZKnWCt07BW6oRt1JoT79C+HzCucYjgz5i2ul1D1UZg44oPkmyqqg0r3X6WrHXHWyt1grVOw1qpE4apdYihns3A/Ucu3w/49gB1SFKThgj+y4ADkuyfZFfgecAZA9QhSU2a+VBPVd2a5I+ATwI7AydV1bVTONSKh4kGYK073lqpE6x1GtZKnTBAram63fC6JGkd85O7ktQYg1+SGrOmgj/JSUluTHLNSNveSc5O8uX+914j1726nxbi+iT/ZaT915N8ob/ur5OMe4vpTOpMMp/kJ0mu7H/eMas6l6j1OUmuTfKLJBsWrT9In25vrUP264Q635jkS0muTvKhJHuOXLfa+nRsrav0vnpCX+eVSc5Kcp+R61bT///YOgfr06paMz/A44FfA64ZafvfwHH98nHAif3yw4CrgDsD+wP/D9i5v+5S4CC6zxT8E/DUAeucH11v0X6mWucStT4UeDDwaWDDSPtgfbqCWgfr1wl1PgXYpV8+cTXcT1dQ62q8r+4xsnws8I6h+3U76xykT9fUGX9VXQD8YFHz4cC7++V3A78z0v6eqvppVX0N+Arw6CT3pvsjXFRd7/79yDZD1DnWLOqE8bVW1XVVNe6T0oP16QpqHWuov39VnVVVt/YXL6b7/Aqszj6dVOtYA9f6w5GLd2Prh0FX1f//EnWONe0611TwT7BvVX0HoP99r7593NQQ9+1/No9pH6pOgP2TfD7J+Un+c982VJ1LWW19ui2rtV9fRHcGt1DPau7T0VphFfZpktcn+SbwfOC1IzWtqn6dUCcM0KfrIfgnmTQ1xLKmjJih7wD7VdUjgVcB/5hkD1ZfnbB2+hRWab8meQ1wK3DqQtOEegbv0zG1rso+rarXVNX96er8o7551fXrhDoH6dP1EPzf658WLTw9urFvnzQ1xGZu+9R1VlNGjK2zfyr6/X75crqxyP80YJ1LWW19OtFq7NckRwNPB57fP32HVdqn42pdjX26yD8Cz+6XV2W/9n5Z51B9uh6C/wzg6H75aOAjI+3PS3LnJPsDBwCX9sMsNyd5bP8q+QtHtpl5nUnm0n1HAUke2Nf51QHrXMpq69OJVlu/JjkM+HPgGVX145GrVl2fTqp1tfVpX8cBIxefAXypX15V/TqpzsH6dEe9SjyLH+A0uqdGP6N7RHwxcE/gHODL/e+9R9Z/Dd0j6PWMvCIObACu6a97G/0nmIeok+6R/1q6dyBcAfz2rOpcotZn9ss/Bb4HfHLoPt3eWofs1wl1foVuzPnK/ucdq7hPx9a6Su+rH+iPezXwUeC+Q/fr9tQ5VJ86ZYMkNWY9DPVIkraDwS9JjTH4JakxBr8kNcbgl6TGGPwzkORHU97/K5LstiOO17/v+VP9TIFH7pgKb3eMqfbHNCS5T5LT++UDk/zWyHXHJ/mTbWx/Sd+n30iyZWQ2xvkkuyS5Kcn/mrDtVUlO27G3CJLsmeSlI5efkORjO/gYt+mrIa3F+920GPzrwyuA3ba10jI9ErhTVR1YVe/dQftc86rq21V1RH/xQGC7wqyqHlNVB9LN0fLevn8PrKob6GbDvB54bv9hnV9K8lC6/9PHJ7nbHbsVt7Mn8NJtrXQHHch29pWmz+AfSJIHJflEksuTXJjkIX37yenm3v5ckq8mOaJv3ynJ36Sbe/5jST6e5IgkxwL3Ac5Lct7I/l/fnylenGTfMcffO8mH080RfnGSRyS5F/APwIH92eiDllnzb/dntJ/vny3s27fvnuTv0s0pfnWSZ4/sa1v1zaX73oIrkrwzydeT7NNf94Ikl/Y1vnPkk48/Grfffl8fSHJZ/3PwmON9PMkj+uXPJ3ltv3xCkj/oz8yvSbIr8D+BIxc9K3pYkk/3f7Njl3Un2Ooo4K+AbwCPXXTdfwVOAc6i+8Tn7SzR/78x8szi80nuvmjTNwAP6q9/Y9+2e5LT083Hf+rCA1G6ueHP7//2n0w//ciiOp7T99FVSS4Y11fj7nf9tscnOSXJuem+s+IlE27rC/ttr0pySt/2gCTn9O3nJNmvb98/yUX93/yEkX0k3XcOXNPfN4/s25/Q/w1vd/vXnR396Tp/xn6S70dj2s4BDuiXHwOc2y+fDLyf7kH5YcBX+vYjgI/37f8B+BfgiP66G4B9RvZd9J8ApPsegL8Yc/y3Aq/rl58IXNkvPwH42ITbManmvdj6/c1/APyffvlE4C0j2++1HfW9DXh1v3xYv80+dPPvf5TuWQnA3wAvXGq/dHOjHNIv7wdcN+Z4xwEvA/YALmPrJ4DPo5vvf55+3nTg94C3jWx7PPA5urnf9wG+v1DfmOMs3vaudHOw7AYcA/z1ovX/GXgA3bOCMybsc1L/fxQ4uF/enX6O/ZHtfnmbRv72/0Y3L8xOwEXAIcCd+ts31693JHDSmDq+wNZPpO454fZOut8dT/fp1bv2ffhN4D6L9v8rdM+M9ukv7z1yO4/ul18EfLhfPmPkvvEy+v9Duk/Lng3sDOxL94B770m3f+j8mMbPLmjmkuwOPA54/8gJxZ1HVvlwVf0C+OLI2fAhwPv79u9m5Ox+jFuAhbHay4Enj1nnELZOFHVuknsmuccKa74f8N7+LHBX4Gt9+5OA5y2sXFX/sp31PbPf7hNJFrY9FPh14LK+jruydWK+Sft9Et0Z+cK+90hy96q6eeR4F9J9QcbXgDOBJ6d73WS+qq5PMj+2Y7Y6s6p+Cvw0yY10gbJ5G9tANxHaeVX14yQfAP57kldW1c+TPArYUlVfT7IZOCnJXiP9uGBS/38WeHOSU4EPVtVy6rl0Yb0kV9I9OPwr8HDg7L4Pd6abkmCxzwInJ3kf8MEJ+1/qfveRqvoJ8JP+/v1o4MMj2z4ROL2qbuq3X5jz/iDgWf3yKXQP+gAHs3XStlPoTkQWajitqn5ON3ni+cCjgB9OuP2fmXBb1iyDfxg7Af9a3ZjvOD8dWc6i38vxs+pPbYCfM/7vvL3Tvi5V81uBN1fVGUmeQHf2tnCMcftcaX0L7e+uqldvx353Ag7qQ2WSy+jmRvkq3dngPsBL6B5AlmP0bzbpNo1zFHBwkhv6y/cEfhP4VH/dQ0au24MuyN61aB9j+7+q3pDkTLox9ouTPKmqvsTSxt2OANdW1UFLbVhVf5jkMcDTgCuTHDhmtaXud4vvK4svT7o/TdrfuH1MqmHBSv+Oa4pj/AOo7tt4vpbkOfDLMcdf3cZmnwGenW6sf1+6p6ULbgYWj99uywV0XwhBHxY31W2/JWh7ar4H8K1++eiRzc5i67zjZOT7kJfhM8Bz++2eQjecAd1w0xHpXo9YeK3iAdvY1+I6Dly8QlXdQje88Fy6b526EPiT/vdiK+nv20k37/ohdPOxz1fVPN2QxFFJdgKeAzxi5LrD6R4MFhvb/0keVFVfqKoTgU3AQ1Z4O64H5pIc1O/3Tkl+ZczteVBVXVJVrwVuopsWefExlrrfHZ7kLknuSXf/vmzRIc6hewH8nv32e/ftn2PrM8vns/UM/bOL2kdrODLJzknm6L4q8dJl9MO6YfDPxm5JNo/8vIrujvjiJFfRzc53+Db28QG6oYNrgHcCl9CNRwJsBP5pG8M/ix0PbEhyNd2LfEcvvTosUfPxdENAF9L9wy/4S2Cv/kW0q+jOZJfrfwBPSXIF8FS6oYWbq+qLwF8AZ/W1n003PruUY+lva5IvAn84Yb0Lge9VNxXxhXRDKOOC/zy6oaM7+pbXZ9G9TjJ6lvkRuhdxnwx8q6q+NXLdBf1xF9/e4xnf/68Y6fufcNtv0qK6eeA/26/zRiboHxSPAE7s93Ul3bDfYm/sXyy9pq/1Km7fV8cz+X53Kd0w28XACVV1m/nnq+pa4PXA+X0db+6vOhb4/X6fvwu8vG9/OfCyJJfRPTgu+BDdLJlXAecCf1ZV3510+9cjZ+dcQ5LsXlU/6s94LqV74W5d3mGT3Bn4eVXd2p9pvn2JoTGtcUmOp3vx9U1D19KCdTl+tY59LMmedC/gnbBeQ7+3H/C+fsjjFrrxdkk7gGf8ktQYx/glqTEGvyQ1xuCXpMYY/JLUGINfkhrz/wGW3vM9ZlzfNgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "x_taa = taa_gene_length\n",
    "taa_bins = 5\n",
    "plt.hist(x_taa, bins=taa_bins)\n",
    "plt.xlabel('Length of each gene with TAA as the stop codon')\n",
    "plt.ylabel('Count')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that most possible genes are 1300-1350 bp long. This is about 10 times lower than the typical gene length in cultivated soybean genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Count')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAWz0lEQVR4nO3dfZRlVXnn8e+PdwUVCE2vVnRaEmKCmRGlNSKaKIghagQzCDIx9mTQnjGOio4xzZg4GWeyxJgXX+LS6aCxJ6JiUISIEbEFNWsUaRQRBIQQVMYO3Sq+oLMk6DN/nF1yqa6urm7qVHX1/n7Wuuues8/bc/ateu6++96zT6oKSVI/9ljsACRJC8vEL0mdMfFLUmdM/JLUGRO/JHVmr8UOYC4OOeSQWrly5WKHIUlLylVXXfXNqlo2vXxJJP6VK1eycePGxQ5DkpaUJF+dqdyuHknqjIlfkjpj4pekzpj4JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SerMkrhyVxrTyrUXz7r81rOfsUCRSAvDFr8kdcbEL0mdMfFLUmdM/JLUGRO/JHVm1MSf5MAk5ye5Icn1SY5JcnCSS5Pc1J4PGjMGSdK9jd3ifxPw0ar6BeBRwPXAWmBDVR0BbGjzkqQFMlriT/JA4FeAdwBU1V1V9R3gJGB9W209cPJYMUiStjZmi/9wYAvw10m+kOScJPsDy6tqE0B7PnTEGCRJ04yZ+PcCHgO8raoeDfyAHejWSbImycYkG7ds2TJWjJLUnTET/23AbVV1RZs/n+GN4PYkKwDa8+aZNq6qdVW1qqpWLVu21U3iJUk7abTEX1X/DHw9ySNa0fHAl4GLgNWtbDVw4VgxSJK2NvYgbS8Bzk2yD3AL8DsMbzbvT3IG8DXgOSPHIEmaMGrir6qrgVUzLDp+zONKkrbNK3clqTMmfknqjIlfkjpj4pekzpj4JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SeqMiV+SOmPil6TOmPglqTMmfknqjIlfkjpj4pekzpj4JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SeqMiV+SOmPil6TO7DXmzpPcCnwf+DFwd1WtSnIwcB6wErgVOLWq7hgzDknSPRaixf+Uqjqqqla1+bXAhqo6AtjQ5iVJC2QxunpOAta36fXAyYsQgyR1a+zEX8DHklyVZE0rW15VmwDa86EzbZhkTZKNSTZu2bJl5DAlqR+j9vEDx1bVN5IcClya5Ia5blhV64B1AKtWraqxApSk3oza4q+qb7TnzcAFwOOA25OsAGjPm8eMQZJ0b6Ml/iT7J3nA1DTwNOBa4CJgdVttNXDhWDFIkrY2ZlfPcuCCJFPHeU9VfTTJlcD7k5wBfA14zogxSNoJK9devM1lt579jAWMRGMYLfFX1S3Ao2Yo/xZw/FjHlSTNzit3JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SeqMiV+SOmPil6TOmPglqTMmfknqjIlfkjoz9nj8knYzsw3gBg7ithTY4pekzpj4JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SeqMiV+SOmPil6TOmPglqTMmfknqjIlfkjozeuJPsmeSLyT5cJs/OMmlSW5qzweNHYMk6R4L0eJ/GXD9xPxaYENVHQFsaPOSpAUyauJPchjwDOCcieKTgPVtej1w8pgxSJLubewW/xuBVwE/mShbXlWbANrzoTNtmGRNko1JNm7ZsmXkMCWpH6Ml/iTPBDZX1VU7s31VrauqVVW1atmyZfMcnST1a8w7cB0LPCvJ04H9gAcmeTdwe5IVVbUpyQpg84gxSJKmGa3FX1VnVdVhVbUSeC7wiap6HnARsLqtthq4cKwYJElbW4zf8Z8NnJDkJuCENi9JWiALcrP1qrocuLxNfws4fiGOK0namlfuSlJnTPyS1BkTvyR1xsQvSZ0x8UtSZ+aU+JMcO5cySdKub64t/rfMsUyStIub9Xf8SY4BngAsS/KKiUUPBPYcMzBJ0ji2dwHXPsABbb0HTJR/DzhlrKAkSeOZNfFX1SeBTyZ5V1V9dYFikiSNaK5DNuybZB2wcnKbqjpujKAkSeOZa+L/W+DtDHfS+vF44UiSxjbXxH93Vb1t1EgkSQtirj/n/Lskv5tkRZKDpx6jRiZJGsVcW/xTN075vYmyAg6f33AkSWObU+KvqoePHYgkaWHMKfEnef5M5VX1v+c3HEnS2Oba1fPYien9GO6g9XnAxC9JS8xcu3peMjmf5EHA34wSkSRpVDs7LPMPgSPmMxBJ0sKYax//3zH8igeGwdl+EXj/WEFJksYz1z7+P52Yvhv4alXdNkI8kqSRzamrpw3WdgPDCJ0HAXeNGZQkaTxzvQPXqcDngOcApwJXJHFYZklaguba1fNq4LFVtRkgyTLg48D529ogyX7Ap4B923HOr6r/1oZ6OI9hpM9bgVOr6o6dPQFJ0o6Z66969phK+s235rDtj4DjqupRwFHAiUkeD6wFNlTVEcCGNi9JWiBzbfF/NMklwHvb/GnAR2bboKoKuLPN7t0eBZwEPLmVrwcuB35/zhFLku6T7d1z9+eA5VX1e0l+E3giEOAzwLnb23mSPYGrgJ8D3lpVVyRZXlWbAKpqU5JDt7HtGmANwMMe9rAdOCVJ0my2113zRuD7AFX1wap6RVW9nKG1/8bt7byqflxVRwGHAY9L8ktzDayq1lXVqqpatWzZsrluJknaju0l/pVVdc30wqrayPDl7JxU1XcYunROBG5PsgKgPW/e9paSpPm2vcS/3yzL7jfbhkmWJTmwTd8PeCrDtQAXcc/4/quBC+cUqSRpXmwv8V+Z5IXTC5OcwdB3P5sVwGVJrgGuBC6tqg8DZwMnJLkJOKHNS5IWyPZ+1XMmcEGS3+KeRL8K2Ad49mwbti6iR89Q/i2GYZ0lSYtg1sRfVbcDT0jyFGDqi9mLq+oTo0cmSRrFXMfjvwy4bORYJO0GVq69eJvLbj37GQsYibZlZ8fjlyQtUSZ+SeqMiV+SOmPil6TOmPglqTMmfknqjIlfkjpj4pekzpj4JakzJn5J6oyJX5I6Y+KXpM6Y+CWpMyZ+SeqMiV+SOjOn8fgljcOx67UYbPFLUmdM/JLUGbt6pO2wO0a7G1v8ktQZE78kdcbEL0mdGS3xJ3loksuSXJ/kuiQva+UHJ7k0yU3t+aCxYpAkbW3MFv/dwH+pql8EHg+8OMmRwFpgQ1UdAWxo85KkBTJa4q+qTVX1+Tb9feB64CHAScD6ttp64OSxYpAkbW1B+viTrAQeDVwBLK+qTTC8OQCHbmObNUk2Jtm4ZcuWhQhTkroweuJPcgDwAeDMqvreXLerqnVVtaqqVi1btmy8ACWpM6Mm/iR7MyT9c6vqg6349iQr2vIVwOYxY5Ak3duYv+oJ8A7g+qr684lFFwGr2/Rq4MKxYpAkbW3MIRuOBX4b+FKSq1vZfwXOBt6f5Azga8BzRoxBkjTNaIm/qv4ByDYWHz/WcSVJs/PKXUnqjKNzSruo2UYFBUcG1c6zxS9JnTHxS1Jn7OpZJH6M12La3t+fdm+2+CWpMyZ+SeqMXT3SfWCXnZYiW/yS1BkTvyR1xsQvSZ0x8UtSZ0z8ktQZE78kdcafc0paMmb7+exi/nR2V41rW2zxS1JnTPyS1Bm7ekbkQFi7jt5ei97OVzvGFr8kdcbEL0mdsatHS4KDoUnzxxa/JHXGxC9JnTHxS1JnRkv8Sd6ZZHOSayfKDk5yaZKb2vNBYx1fkjSzMVv87wJOnFa2FthQVUcAG9q8JGkBjZb4q+pTwLenFZ8ErG/T64GTxzq+JGlmC93Hv7yqNgG050O3tWKSNUk2Jtm4ZcuWBQtQknZ3u+yXu1W1rqpWVdWqZcuWLXY4krTbWOjEf3uSFQDtefMCH1+SurfQif8iYHWbXg1cuMDHl6TujflzzvcCnwEekeS2JGcAZwMnJLkJOKHNS5IW0Ghj9VTV6dtYdPxYx5Qkbd8u++WuJGkcjs65i1pq9/CUFpsjuM6dLX5J6oyJX5I6031Xz+54b1I/8u46dse/Ly19tvglqTMmfknqzG7f1eNHbUm6N1v8ktQZE78kdWa37+rRwvHXRAtrKXZj+jeytcWoE1v8ktQZE78kdcbEL0mdsY9fu4Wl2N8tLRZb/JLUGRO/JHXGrh5Ju4zF6rIb87i7YjekLX5J6oyJX5I6Y1fPErQrfnScC6/a1GJaqv83Y7DFL0mdMfFLUmdM/JLUmUXp409yIvAmYE/gnKo6ezHi0K7FPlhpYSx4iz/JnsBbgV8HjgROT3LkQschSb1ajK6exwE3V9UtVXUX8D7gpEWIQ5K6tBhdPQ8Bvj4xfxvwy9NXSrIGWNNm70xy4wLENoZDgG8udhCT8voFP+QuVweLxHoYWA+DOdXDffx//VczFS5G4s8MZbVVQdU6YN344YwrycaqWrXYcSwm62BgPQysh8Fi1sNidPXcBjx0Yv4w4BuLEIckdWkxEv+VwBFJHp5kH+C5wEWLEIckdWnBu3qq6u4k/xm4hOHnnO+squsWOo4FtOS7q+aBdTCwHgbWw2DR6iFVW3WvS5J2Y165K0mdMfFLUmdM/DsoyUOTXJbk+iTXJXlZKz84yaVJbmrPB01sc1aSm5PcmOTXJsqPTvKltuzNSWb6qesuK8meSb6Q5MNtvsc6ODDJ+UluaH8Tx3RaDy9v/w/XJnlvkv16qIck70yyOcm1E2Xzdt5J9k1yXiu/IsnKeQm8qnzswANYATymTT8A+ArD0BN/Aqxt5WuB17fpI4EvAvsCDwf+EdizLfsccAzDtQ1/D/z6Yp/fDtbFK4D3AB9u8z3WwXrgBW16H+DA3uqB4aLMfwLu1+bfD/z7HuoB+BXgMcC1E2Xzdt7A7wJvb9PPBc6bl7gXu+KW+gO4EDgBuBFY0cpWADe26bOAsybWv6S9wCuAGybKTwf+12Kfzw6c92HABuC4icTfWx08sCW8TCvvrR6mrsY/mOGXgh8GntZLPQArpyX+eTvvqXXa9F4MV/rmvsZsV8990D52PRq4AlheVZsA2vOhbbWZhqh4SHvcNkP5UvFG4FXATybKequDw4EtwF+3Lq9zkuxPZ/VQVf8X+FPga8Am4LtV9TE6q4cJ83neP92mqu4Gvgv8zH0N0MS/k5IcAHwAOLOqvjfbqjOU1Szlu7wkzwQ2V9VVc91khrIlXQfNXgwf899WVY8GfsDw0X5bdst6aH3YJzF0XzwY2D/J82bbZIayJV8Pc7Az5z1KnZj4d0KSvRmS/rlV9cFWfHuSFW35CmBzK9/WEBW3tenp5UvBscCzktzKMLrqcUneTV91AEP8t1XVFW3+fIY3gt7q4anAP1XVlqr6F+CDwBPorx6mzOd5/3SbJHsBDwK+fV8DNPHvoPZt+zuA66vqzycWXQSsbtOrGfr+p8qf276dfzhwBPC59hHw+0ke3/b5/IltdmlVdVZVHVZVKxm+cPpEVT2PjuoAoKr+Gfh6kke0ouOBL9NZPTB08Tw+yf1b/McD19NfPUyZz/Oe3NcpDP9r9/1T0GJ/MbLUHsATGT5qXQNc3R5PZ+h32wDc1J4Pntjm1Qzf4N/IxK8UgFXAtW3ZXzIPX9osQn08mXu+3O2uDoCjgI3t7+FDwEGd1sN/B25o5/A3DL9c2e3rAXgvw/ca/8LQOj9jPs8b2A/4W+Bmhl/+HD4fcTtkgyR1xq4eSeqMiV+SOmPil6TOmPglqTMmfknqjIm/M0mWJ3lPkluSXJXkM0mevdhxTZfkwUnOb9NHJXn6xLI/SvLK7Wx/RZKrk3wtyZY2fXWSlUn2SvLNJK+bts0BSd6W5B/bEAxXJXnhPJ/XyUmOnJi/PMm83nA7yZlJ7j+f+9zJOLb7OmlxmPg70i4O+RDwqao6vKqOZrgA67BZN1wEVfWNqjqlzR7FcK3Ejmz/y1V1FPAahhENj2qPW7lnALFTpw37ew5wB3BEDUMwnMgw8Nh8OplhlMYxnQkseuLXrsvE35fjgLuq6u1TBVX11ap6C/x0fP03JLkyyTVJ/mMrf3JrmU6NO3/uxHjhRyf5ZGsdXzJxqfpLk3y57ed90wNJ8pEk/6ZNfyHJa9r0/0jygtYyvzbJPsBrgdNai/20tosjW0y3JHnpDtbD6cCbaFectuP+LPA44A+q6ietbrZU1etn2kGSD7Vzvi7Jmon6e1eL+0tJXj5tmycAzwLe0M7lZ9ui5yT5XJKvJHnSbK/FtP3tn+TiJF9sxzyt1cWDgcuSXNbWO73Fc22S109sf2eSP0vy+SQbkiyb4RjLk1zQjvHFdg4keUXb37VJzpxY/9UZxpr/OPCIifKjkny2ncsFaWPUt9fw9dPPXyNb7CvffCzcA3gp8BezLF/DkPhguPJyI8PAW09mGBXwMIbGwmcYrmDeG/g/wLK2zWnAO9v0N4B92/SBMxxrLfBihqGNrwQuaeWXMSSMlbShbhnGdv/LiW3/qB13X+AQ4FvA3ts4p+nb3q/Fdv92vm9u5c8CLtiBujx4Yn/XMlyteTRw6cQ6M533u4BTJuYvB/6sTT8d+Phsr8W0ff1b4K8m5h/Unm8FDmnTD2Z4g1vGMKjcJ4CT27ICfqtNv2aynib2eR7DQIQAezKMFXM08CVgf+AA4DqGUWqnyu/fXtebgVe2ba8BfrVNvxZ442zn72Pchy3+jiV5a2vFXdmKngY8P8nVDENN/wzDeCIwjClyWw2t4asZEvMjgF8CLm3b/AH3dBtdA5ybYZTGu2c4/KcZbmLxROBi4IDWL72yqm6cQ/gXV9WPquqbDINgLZ/jaT8TuKyqfsgw0N6zk+w5faXWcr06ybYGCXtpki8Cn2UYROsI4Bbg8CRvSXIiMNuorZOmBvq7iqFeYfbXYsqXgKe2FvOTquq7M+z7scDlNXx6uRs4l6HeYRhS+7w2/W6G12K644C3AVTVj9sxnsjwJvmDqrqzxf+k9rigqn5Yw4i1FwEkeRDDm+An2z7XT8SwrfPXiPZa7AC0oK5jaCUCUFUvTnIIQ2sShiFgX1JVl0xulOTJwI8min7M8LcT4LqqOmaGYz2D4Z/7WcAfJnlkSzxTrmQYn+QW4FKGlvsLGf7552KmeObidODYDCOLwpBQn8IwuNqjkuxRVT+pqj8G/jjJndN30OrjqQw3yPhhksuB/arqjiSPAn6N4dPMqcB/2IFzmTyPGV+LSVX1lSRHM7SUX5fkY1X12unhzuH4P93lHNebbZ87MwbMTOevEdni78sngP2SvGiibPJLwEuAF2UYdpokP5/hxiLbciOwLMkxbf29kzwyyR7AQ6vqMoabtRzI0CXwU1V1F8MNJk5laDV/Gnhle57u+wy3ubxPkjyQobX6sKpaWcPooi8GTq+qmxneAP/n1CeAJPsxc5J7EHBHS/q/wD3fExwC7FFVHwD+kGGI5p09l+2+FkkeDPywqt7NcCOUqeNNHuMK4FeTHNLO63RgquW9B8OIjwD/DviHGeLYALyoHW/PVoefAk7OMBrn/sCzGV63TzF8grpfkgcAvwHQPiXcMdF//9sTMWgR+O7akaqqJCcDf5HkVQx3j/oB8PttlXMYPmp/Pkna8pNn2d9dSU4B3tw+zu/FcGeurwDvbmVh+F7hOzPs4tPA8S2Bfpqhm2imxH8ZsLZ1e7xuhuVz9ZsMw9pOflq4EPiTJPsCLwDeANyc5NvA/+Oeupn0UeA/JbmG4c3vs638IQx345pqUJ01w7bvA/6qfQl7ygzLp8zltfjXDF8U/4RhdMipN/R1wN8n2VRVT0lyFkMdBvhIVU0N+fsD4JFJrmL4Duc0tvYyYF2SMxha5C+qqs8keRfDaJEA51TVFwCSnMfQFfhV7v1argbe3rrzbgF+Z5Zz18gcnVPqVJI7q+qA7a+p3Y1dPZLUGVv8ktQZW/yS1BkTvyR1xsQvSZ0x8UtSZ0z8ktSZ/w+axhdDmaBU7QAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# repeating for TAG stop codon\n",
    "\n",
    "x_tag = tag_gene_length\n",
    "tag_bins = 40\n",
    "plt.hist(x_tag, bins=tag_bins)\n",
    "plt.xlabel('Genes with TAG as the stop codon')\n",
    "plt.ylabel('Count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With TAG as the stop codon, we see two peaks. One peak is around 5500 bp length, the other peak is around 7500 bp gene length. These are looking more like they could be actual genes (though further analysis needs to be done to confirm this), because their nucleotide length is closer to the typical gene length for cultivated soybean of 18,000."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### I worked on the 1037-40810 (1036-40810 using Python indexing) range of alignment for this project. Taking the TATA box that preceeds every gene and the FASTA file header length into account, I ofset the starting index by the length of the header and the first index of the TATA box. I found indices of candidate start and stop codons and used them to find candidate genes. I found 25 candidate genes for the TAA (UAA in RNA) stop codon and 477 for the TAG (UAG in RNA) stop codon."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
